File: NCoreUtils.AspNetCore.Rest/DefaultRestDelete.fs
namespace NCoreUtils.AspNetCore.Rest

open System.Runtime.CompilerServices
open Microsoft.Extensions.Logging
open NCoreUtils.AspNetCore
open NCoreUtils.Data
open NCoreUtils.Logging

/// <summary>
/// Provides default implementation for REST DELETE method.
/// </summary>
/// <typeparam name="a">Type of the target object.</typeparam>
/// <typeparam name="id">Type of the Id property of the target object.</typeparam>
type DefaultRestDelete<'a, 'id when 'a :> IHasId<'id> and 'id : equality> =
  inherit DefaultTransactedMethod<'a, 'id>
  val private logger : ILogger

  /// <summary>
  /// Initializes new instance from the specified parameters.
  /// </summary>
  /// <param name="repository">Repository to use.</param>
  /// <param name="loggerFactory">Logger factory.</param>
  [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
  new (repository, loggerFactory : ILoggerFactory) =
    { inherit DefaultTransactedMethod<'a, 'id> (repository)
      logger = loggerFactory.CreateLogger "NCoreUtils.AspNetCore.Rest.DefaultRestDelete" }

  abstract AsyncInvoke : id:'id -> Async<unit>

  /// <summary>
  /// Performes REST DELETE action for the predefined type.
  /// </summary>
  /// <param name="id">Id of the object to delete.</param>
  default this.AsyncInvoke id = async {
    let! item = this.Repository.AsyncLookup id
    match box item with
    | null ->
      debugf this.logger "No entity of type %s found for key = %A (data-delete)." typeof<'a>.FullName id
      NotFoundException () |> raise
    | _ ->
      do! this.Repository.AsyncRemove item
      debugf this.logger "Successfully removed entity of type %s with key = %A (data-delete)." typeof<'a>.FullName id }

  interface IRestDelete<'a, 'id> with
    member this.AsyncInvoke id = this.AsyncInvoke id
  interface IBoxedInvoke<'id, unit> with
    member this.Instance = box this
    member this.AsyncInvoke id = this.AsyncInvoke id

File: NCoreUtils.AspNetCore.Rest/Rest/DefaultRestDelete.cs
using System;
using System.Diagnostics.CodeAnalysis;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using NCoreUtils.Data;

namespace NCoreUtils.AspNetCore.Rest;

/// <summary>
/// Provides default implementation for REST DELETE method.
/// </summary>
/// <typeparam name="TData">Type of the target object.</typeparam>
/// <typeparam name="TId">Type of the Id property of the target object.</typeparam>
/// <remarks>
/// Initializes new instance from the specified parameters.
/// </remarks>
/// <param name="repository">Repository to use.</param>
/// <param name="logger">Logger to use.</param>
public class DefaultRestDelete<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties)] TData, TId>(
    IDataRepository<TData, TId> repository,
    ILogger<DefaultRestDelete<TData, TId>> logger)
    : DefaultTransactedMethod<TData, TId>(repository)
    , IRestDelete<TData, TId>
    , IBoxedVoidInvoke<IRestDeleteContext<TData, TId>>
    where TData : IHasId<TId>
{
    protected ILogger Logger { get; } = logger ?? throw new ArgumentNullException(nameof(logger));

    object IBoxedInvoke.Instance => this;

    /// <summary>
    /// Performes REST DELETE action for the predefined type.
    /// </summary>
    /// <param name="context">Invocation context.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    public virtual async ValueTask InvokeAsync(IRestDeleteContext<TData, TId> context, CancellationToken cancellationToken)
    {
        var id = context.Id;
        var item = await Repository.LookupAsync(id, cancellationToken);
        if (item is null)
        {
            Logger.LogRestNoEntityFound(typeof(TData), id);
            throw new NotFoundException();
        }
        await Repository.RemoveAsync(item, context.Force, cancellationToken: cancellationToken);
        Logger.LogRestEntityRemovedSuccessfully(typeof(TData), id);
    }
}